dali/python/nvidia/dali/external_source.py

```diff
@@ -296,13 +330,28 @@ class ExternalSource():
     of separate Tensors, there will be an additional copy made internally, consuming both memory
     and bandwidth.
 
+    Automatically set to ``True`` when ``parallel=True``
+
 `batch` : bool, optional
     If set to ``True`` or ``None``, the ``source`` is expected to produce an entire batch at once.
     If set to ``False``, the ``source`` is called per-sample.
+
+`parallel` : bool, optional, default = False
+    If set to ``True`` the corresponding pipeline will run pool of Python workers to run the
+    callback in parallel. You can specify the number of workers by passing ``py_workers_num``
+    into pipeline's constructor.
+
+    When parallel=True ``source`` must return NumPy/MXNet/PyTorch CPU array, TensorCPU, or
```

Comment:
Added support of more types (MXNet, Pytorch, TensorCPU), but without array_interface.

It basically works by converting those types to numpy array. At least in case of pytorch it doesn't introduce any significant overload because there's no copy involved. I wanted to use TensorCPU as this common type of conversion (which would remove numpy dependency) but for now it seems to lack a way of checking out what number of bytes its copy_to_external method would write to the buffer, but most importantly its constructor doesn't provide a way to hint it with the info about shape and type of the array from the buffer.

array_interface as a type of "meta-data" to pass between processes doesn't seem to be exactly what's needed because it uses pointers and what we need is offsets (as shared memory will most probably be mapped to different addresses in the parent and the worker.

If it was possible to construct TensorCPU with array_interface we could use simple serialization that only changes addresses to offset in the worker and offsets to new addresses in the parent, eliminating the need for numpy.